File: moderate/reverse_groups.cs
/* CodeEval - Reverse Groups
 * https://www.codeeval.com/public_sc/71/ 
 */

using System.IO;
using System.Collections.Generic;

class Program
{
    static void ReverseArrayParts(string[] array, int x, int y)
    {
        for (int i = x; i <= (x+y)/2; i++){
            int j = i-x;
            string a = array[i];
            string b = array[y-j];
            array[i] = b;
            array[y-j] = a;
        }
    }

    static void ReverseGroup(string[] array, int k)
    {
        int max = array.Length/k * k;        
        int i = 0;
        int j = k -1;
        while (j < max)
        {
            ReverseArrayParts(array, i, j);
            i = j + 1;
            j = i + k - 1;
        }
    }

    static void Main(string[] args)
    {
        using (StreamReader reader = File.OpenText(args[0]))
        while (!reader.EndOfStream)
        {
            string line = reader.ReadLine();
            if (null == line)
                continue;
            string[] tokens = line.Trim().Split(';');
            int k = int.Parse(tokens[1]);
            string[] numberArray = tokens[0].Split(',');
            ReverseGroup(numberArray, k);
            string output = string.Join("," , numberArray);
            System.Console.WriteLine(output);
        }
    }
}

File: moderate/reverse_groups.py
#!/usr/bin/env python2
# CodeEval - Reverse Groups
# https://www.codeeval.com/public_sc/71/

from sys import argv

def reverse_group(number_list, k):
    new_list = number_list[:]
    max_ = int(len(new_list)/k) * k 
    i = 0
    j = k
    while j <= max_:
        new_list = new_list[:i] + new_list[i:j][::-1] + new_list[j:]
        i = j 
        j += k
    return new_list


def main():
    f = open(argv[1], "rt")
    for line in f:
        number_list, k = line.strip().split(";")
        k = int(k)
        number_list = number_list.split(",")
        new_list = reverse_group(number_list, k) 
        print( ",".join(new_list) )
    f.close()


if __name__ == "__main__":
    main()
